content/rust-compiler-performance-survey-2025-results/index.md

```diff
@@ -0,0 +1,325 @@
+
+Approximately 60% of respondents say that they use `cargo` terminal commands to type check, build or test their code, with `cargo check` being the most commonly used command performed after each code change:
+
+<!-- Chart cargo-commands start -->
+<div>
+    <div class="matrix-chart" id="cargo-commands" style="height:500px; width:100%;"><noscript>
+<img alt="cargo-commands" height="500" src="cargo-commands.png"/>
+</noscript></div>
+    <div style="display: flex; margin-bottom: 10px;">
+        <span>[<a href="cargo-commands.png" target="_href_" title="Download chart as PNG">PNG</a>]</span>&nbsp;<span>[<a href="cargo-commands.svg" target="_href_" title="Download chart as SVG">SVG</a>]</span>
+    </div>
+</div>
+<!-- Chart cargo-commands end -->
+
+While the performance of `cargo check` does not seem to be as big of a blocker as e.g. incremental rebuilds, it also causes some pain points. One of the most common ones present in the survey responses is the fact that `cargo check` does not share the build cache with `cargo build`. This causes additional compilation to happen when you run e.g. `cargo check` several times to find all type errors, and when it succeeds, you follow up with `cargo build` to actually produce a built artifact. This workflow is an example of competing trade-offs, because sharing the build cache between these two commands by unifying them more would likely make `cargo check` itself slightly slower, which might be undesirable to some users. It is possible that we might be able to find some middle ground to improve the status quo though. You can follow updates to this work in [this issue][cargo-check-build-reuse-issue].
+
+A related aspect is the latency of type checking in code editors and IDEs. Around 87% of respondents say that they use inline annotations in their editor as the primary mechanism of inspecting compiler errors, and around 30% of respondents consider waiting for these annotations to be a big blocker. In the open answers, we also received many reports of Rust Analyzer's performance and memory usage being a limiting factor.
+
+The maintainers of Rust Analyzer are working hard on improving its performance. Its caching system is [being improved][ra-talk-rustweek] to reduce analysis latency, the distributed builds of the editor are now [optimized with PGO][ra-pgo], which provided 15-20% performance wins, and work is underway to integrate the compiler's [new trait solver][ra-new-trait-solver] into Rust Analyzer, which could eventually also result in increased performance.
+
+More than 35% users said that they consider the IDE and Cargo blocking one another to be a big problem. There is an existing workaround for this, where you can configure Rust Analyzer to use a different target directory than Cargo, at the cost of increased disk space usage. We realized that this workaround has not been documented in a very visible way, so we added it to the [FAQ section][ra-cargo-blocking] of the Rust Analyzer book.
```

Comment:
If we want to link out to an issue for this, it would be https://github.com/rust-lang/cargo/issues/4282.

The first milestone for this was completed in https://github.com/rust-lang/cargo/issues/14125.  There is work on the next milestone, being tracked at https://github.com/rust-lang/cargo/issues/15010.

(I leave it to you for how much detail to go into or whether to have this at all)